python/ray/ml/trainer.py

```diff
@@ -54,6 +142,69 @@ def __init__(
 
         raise NotImplementedError
 
+    @DeveloperAPI
+    def setup(self) -> None:
+        """Perform initial setup on the Trainer.
+
+        Note: this method is run on a remote process.
+
+        This method will not be called on the driver, so any expensive setup
+        operations should be placed here and not in ``__init__``.
+
+        This method is called prior to ``_preprocess_datasets`` and
+        ``training_loop``.
+        """
+        raise NotImplementedError
+
+    def _preprocess_datasets(self) -> None:
+        """Preprocess dataset attributes with provided preprocessor.
+
+        Note: This method is run on a remote process.
+
+        This method is called prior to entering the training_loop.
+
+        If the ``Trainer`` has both a train_dataset and
+        preprocessor, and the preprocessor has not yet been fit, then it
+        will be fit on the train_dataset.
+
+        Then, the Trainer's train_dataset and any extra_datasets
+        will be transformed by its preprocessor.
+
+        The transformed datasets will be set back in the
+        ``self.train_dataset`` and ``self.extra_datasets`` attributes to be
+        used when overriding ``training_loop``.
+        """
+        raise NotImplementedError
+
+    @DeveloperAPI
+    @abc.abstractmethod
```

Comment:
Why is only this one abstractmethod? Let's make all three similar in annotations.

Reply:
This is the only method that _must_ be overridden for Trainer to work. So, we denote as abstract method to even prevent trainer initialization if this method is not overridden.

Reply:
Ok.